### PR TITLE
feat(invoices): void invoices

### PR DESCRIPTION
**Pull request by ivannovosad**
## Roadmap Task

👉  https://getlago.canny.io/feature-requests/p/ability-to-void-an-invoice

## Context

When an invoice is finalized, it’s currently not possible to delete or void this invoice.

## Description

When an invoice is finalized and generated, add the option to delete or void this invoice through Lago (API or User interface).

### CODE REVIEW

File: app/services/invoices/void_service.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      begin
+        invoice.void!
+
+        invoice.credits.each do |credit|
+          CreditNotes::RecreditService.new(credit:).call
+        end
+
+        invoice.wallet_transactions.each do |wallet_transaction|
+          WalletTransactions::RecreditService.new(wallet_transaction:).call
+        end
```
Comment (ivannovosad) [inline app/services/invoices/void_service.rb:24]: Should we log possible errors for these wallet transactions?

File: app/services/invoices/void_service.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      return result.not_found_failure!(resource: 'invoice') if invoice.nil?
+
+      result.invoice = invoice
+
+      begin
+        invoice.void!
+
+        invoice.credits.each do |credit|
+          CreditNotes::RecreditService.new(credit:).call
+        end
```
Comment (ivannovosad) [inline app/services/invoices/void_service.rb:20]: Should we log possible errors for these wallet transactions?

File: app/services/invoices/void_service.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    def call
+      return result.not_found_failure!(resource: 'invoice') if invoice.nil?
+
+      result.invoice = invoice
+
+      begin
+        invoice.void!
+
+        invoice.credits.each do |credit|
+          CreditNotes::RecreditService.new(credit:).call
```
Comment (vincent-pochet) [inline app/services/invoices/void_service.rb:19]: You can simply use the following form, as all services inherit from the BaseService, which has a `call`  class method that is just initializing an instance with all received parameters and calling the `call` method.

```suggestion
          CreditNotes::RecreditService.call(credit:)
```

**Review by vincent-pochet — COMMENTED** [review]
Looks great!

I added few comments, but no blocker for me. Let me know if you have question about it.

File: app/services/wallet_transactions/recredit_service.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    def call
+      result.wallet_transaction = wallet_transaction
+
+      return result.not_allowed_failure!(code: 'wallet_not_active') unless wallet.active?
+
+      transaction_result = transaction_service.create(
+        organization_id: customer.organization_id,
+        wallet_id: wallet.id,
+        granted_credits: wallet_transaction.credit_amount.to_s,
```
[the remark this answers is not in the document]
Comment (ivannovosad) [reply app/services/wallet_transactions/recredit_service.rb:22]: I don't think so, we're recrediting the same wallet (with the same rate). It's handled [here](https://github.com/getlago/lago-api/blob/main/app/services/wallet_transactions/create_service.rb#L56).

**Review by vincent-pochet — APPROVED** [review]
:rocket:

**Pull request merged**